### ISSUE TITLE
Color Correction

### ISSUE THREAD

**Issue by jakiestfu** [jakiestfu/himawari.js#24]
Provide an optional parameter to color correct the resulting imagery to better reflect images such as the following: 

<a href="https://en.wikipedia.org/wiki/File:Himawari_8_Full_Disk_Aug_21_2015_0210Z.jpg"><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/c/ce/Himawari_8_Full_Disk_Aug_21_2015_0210Z.jpg/1200px-Himawari_8_Full_Disk_Aug_21_2015_0210Z.jpg"></a>

**Comment by ungoldman** [comment]
@jakiestfu what color correction techniques are you using above?

**Comment by jakiestfu** [comment]
@ngoldman From the wiki article:

> Example of a Rayleigh-corrected, true-color full disk image created from the AHI sensor.

**Comment by mdcarter** [comment]
This would be awesome.
Is it doable ?

**Comment by jakiestfu** [comment]
@mdcarter color correction, yes, but Rayleigh Correction, not sure...

**Comment by Joshfindit** [comment]
Just a layman's observation based on reading [Correction of Rayleigh scattering effects in cloud optical thickness retrievals](http://modis-atmos.gsfc.nasa.gov/_docs/Wang%20&%20King%20%281997%29.pdf), in order to correct for Rayleigh scattering (which affects the optical density of clouds), it's necessary to measure the density more directly.
They mention using images taken at different wavelengths, and they also talk about using images taken at different angles, both of which are not available from a single flat image.
As it looks like the source that himawari.js pulls from is a single flat image, calculating Rayleigh scattering correction may not be technologically possible.

**Comment by Joshfindit** [comment]
The question is: what colour correction methods _can_ get close enough to the same result?

**Comment by ungoldman** [comment]
You're not going to be able to get quite the same result as above working with a single full disk image but you could use @celoyd's [color tweaks](https://gist.github.com/celoyd/b92d0de6fae1f18791ef#color-tweaks) suggestions from his Himawari 8 animation tutorial as a starting point. He's using convert which is a part of ImageMagick, so that tool is already available to this project with its current dependencies.

**Comment by jakiestfu** [comment]
@ngoldman do you think you could provide a sample picture or some code that outputs that picture?

**Comment by jakiestfu** [comment]
This code works fine, however, it doesn't do much to the image other than brighten and add contrast:

``` sh
convert original.jpg -channel R -gamma 1.2 -channel G -gamma 1.1 +channel -sigmoidal-contrast 3,50% updated.jpg
```

[original.jpg](http://i.imgur.com/n9yrTUH.jpg), [updated.jpg](http://i.imgur.com/5m07N5U.jpg)

**Comment by celoyd** [comment]
> it doesn't do much to the image other than brighten and add contrast

Right. It tones down blues (and greens) too, but it is extremely simple compared to what [CIRA RAMMB](http://rammb.cira.colostate.edu/ramsdis/online/himawari-8.asp) has done to the image on Wikipedia. Their code is based on what’s used with MODIS and VIIRS, which is [nontrivial](http://modis.gsfc.nasa.gov/data/atbd/atbd_mod08.pdf). They have a paper in review, so more details should be public soon. I’m pretty confident that it’s more complex than anything that makes sense to implement here. (But skip to the “However” heading for some ideas.)

## Rationale for simple adjustment

### We don’t have good access to data for complex adjustment

Our options are limited because we don’t have the bands that CIRA does. They’re able to calculate optical depth, cloud height, etc., from information that’s at best only _kinda_ present in the RGB PNGs we’re looking at. And they can mix some of the NIR channel into the green channel to account for the green band being to the &lt; &lambda; (blue) side of the 550 nm peak visible reflectance of chlorophyll.

Sidebar: _Why? Because the data is produced by the government of Japan, and they haven’t licensed and distributed it that way, as far as I can tell. I’ve signed up for their [P-Tree service](http://www.eorc.jaxa.jp/ptree/), but [its TOS](http://www.eorc.jaxa.jp/ptree/terms.html) is vague and refers to other lengthier and confusinger TOSes. They seem to imagine forecasters, researchers, and resellers as the only potential users – which is completely understandable, but frustrating in our position._

_So while I’m very grateful that they’ve done as much as they have to license and distribute this data – they’ve clearly worked hard to serve their intended users well – I would like to convince them to do a little more. I don’t have the language skills, the contacts, or (at the moment) the time to make a persuasive case for truly open data here. If someone else does, I’d be happy to contribute some sort of amicus brief based on professional experience with these issues. End sidebar._

We just don’t have the raw information that CIRA RAMMB does. If we could get it, it’s not clear (to me, yet) that we could “publish” it.

### We don’t necessarily want complex adjustment

However! CIRA’s path is not necessarily ideal. While I would use the NIR&rarr;green trick if I could, everything else they’re doing is more science- than esthetics- or realism-oriented. A person in space would see Rayleigh scattering making the atmosphere bluer toward the horizon, for example, and [BRDF effects](https://en.wikipedia.org/wiki/Bidirectional_reflectance_distribution_function), both of which the correction deliberately and efficiently removes. They would also see the halo of the atmosphere, which CIRA’s correction cuts out, and they would not see the smudgy artifacts that the correction sometimes introduces near the terminator (dawn and dusk): look west of India in the example image at full size. What CIRA’s doing is extremely impressive, cutting-edge correction. But it’s not necessarily what makes sense here – at least as I’ve been envisioning it.

### However

I’m all for experiments more elaborate adjustment!

> what colour correction methods _can_ get close enough to the same result?

You can model the atmosphere as a spherical shell of known radii (in pixel dimensions) and do some very light trig to find the distance for which a given pixel’s ray intersects it, then use that field to weight a correction. And you can add low-res gridded elevation data ([e.g.](https://www.ngdc.noaa.gov/mgg/global/global.html)) to account for the fact that, for example, the Tibetan Plateau extends above most of the optical atmosphere. Instead of estimating, you can look up measured optical depth, from MODIS or from ground-based weather reports. Or you could pull in something like [6S](https://en.wikipedia.org/wiki/6S_%28radiative_transfer_code%29), which is a standard atmospheric corrector that’s actually related to what CIRA is using. It’s really just a question of how far you want to go.

## So

That’s why I’m pretty happy to use a simple/simplistic static adjustment. For input to `ffmpeg`, incidentally, I’ve been using this:

```
convert -gamma 1.33 -channel B -gamma 0.9 -channel G -gamma 0.975 +channel -sigmoidal-contrast 3,33%
```

Which is a bit light and low-contrast by itself, but whatever random color profiles `ffmpeg` makes up seem to put it about where I want it.

## Just one last point and I’ll give you back the mic

You can [download CIRA’s images](http://rammb.cira.colostate.edu/ramsdis/online/himawari-8.asp) if you prefer them for any reason! There could even be flags to produce, say:
- CIRA
- NICT
- NICT with basic correction
- NICT with experimental correction

## Okay, I’ll be quiet now

**Comment by ungoldman** [comment]
Thanks @celoyd for taking time to get into the details! :grin: :ok_hand: :100:

**Comment by Joshfindit** [comment]
Re: hue correction:

I just ran `convert -gamma 1.33 -channel B -gamma 0.9 -channel G -gamma 0.975 +channel -sigmoidal-contrast 3,33%`, and got the following image: 
![himawariearth post-processed and resized](https://cloud.githubusercontent.com/assets/1937155/13275442/92353006-da8b-11e5-8a6e-4eef08bf0b77.jpg)

The image at the top of this issue has quite a bit more green and Australia is noticeably different.
Is there a more correct colouring?

**Comment by celoyd** [comment]
> The image at the top of this issue has quite a bit more green and Australia is noticeably different.

Yes. This is mainly because CIRA does NIR&rarr;green mixing to account for the green channel not being on the chlorophyll peak.

> Is there a more correct colouring?

Of the two, or theoretically? Could you be more specific?

**Comment by Joshfindit** [comment]
Is there a more correct colouring of the two.
If it's the top image as it seems to be, it would get closer by adding some green and nudging down the red a bit, but that's just off the top of my head.
